/**
 * Autogenerated by Avro
 * 
 * DO NOT EDIT DIRECTLY
 */
package org.exascience.formats.avro;  
@SuppressWarnings("all")
@org.apache.avro.specific.AvroGenerated
public class SnpEffAnnotations extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord {
  public static final org.apache.avro.Schema SCHEMA$ = new org.apache.avro.Schema.Parser().parse("{\"type\":\"record\",\"name\":\"SnpEffAnnotations\",\"namespace\":\"org.exascience.formats.avro\",\"fields\":[{\"name\":\"variant\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Variant\",\"namespace\":\"org.bdgenomics.formats.avro\",\"fields\":[{\"name\":\"contig\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Contig\",\"doc\":\"Record for describing a reference assembly. Not used for storing the contents\\n of said assembly.\\n\\n @see NucleotideContigFragment\",\"fields\":[{\"name\":\"contigName\",\"type\":[\"null\",\"string\"],\"doc\":\"The name of this contig in the assembly (e.g., \\\"chr1\\\").\",\"default\":null},{\"name\":\"contigLength\",\"type\":[\"null\",\"long\"],\"doc\":\"The length of this contig.\",\"default\":null},{\"name\":\"contigMD5\",\"type\":[\"null\",\"string\"],\"doc\":\"The MD5 checksum of the assembly for this contig.\",\"default\":null},{\"name\":\"referenceURL\",\"type\":[\"null\",\"string\"],\"doc\":\"The URL at which this reference assembly can be found.\",\"default\":null},{\"name\":\"assembly\",\"type\":[\"null\",\"string\"],\"doc\":\"The name of this assembly (e.g., \\\"hg19\\\").\",\"default\":null},{\"name\":\"species\",\"type\":[\"null\",\"string\"],\"doc\":\"The species that this assembly is for.\",\"default\":null}]}],\"doc\":\"The reference contig that this variant exists on.\",\"default\":null},{\"name\":\"start\",\"type\":[\"null\",\"long\"],\"doc\":\"The 0-based start position of this variant on the reference contig.\",\"default\":null},{\"name\":\"end\",\"type\":[\"null\",\"long\"],\"doc\":\"The 0-based, exclusive end position of this variant on the reference contig.\",\"default\":null},{\"name\":\"referenceAllele\",\"type\":[\"null\",\"string\"],\"doc\":\"A string describing the reference allele at this site.\",\"default\":null},{\"name\":\"alternateAllele\",\"type\":[\"null\",\"string\"],\"doc\":\"A string describing the variant allele at this site. Should be left null if\\n   the site is a structural variant.\",\"default\":null},{\"name\":\"svAllele\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"StructuralVariant\",\"fields\":[{\"name\":\"type\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"StructuralVariantType\",\"doc\":\"Descriptors for the type of a structural variant. The most specific descriptor\\n should be used, if possible. E.g., duplication should be used instead of\\n insertion if the inserted sequence is not novel. Tandem duplication should\\n be used instead of duplication if the duplication is known to follow the\\n duplicated sequence.\",\"symbols\":[\"DELETION\",\"INSERTION\",\"INVERSION\",\"MOBILE_INSERTION\",\"MOBILE_DELETION\",\"DUPLICATION\",\"TANDEM_DUPLICATION\"]}],\"doc\":\"The type of this structural variant.\",\"default\":null},{\"name\":\"assembly\",\"type\":[\"null\",\"string\"],\"doc\":\"The URL of the FASTA/NucleotideContig assembly for this structural variant,\\n   if one is available.\",\"default\":null},{\"name\":\"precise\",\"type\":[\"boolean\",\"null\"],\"doc\":\"Whether this structural variant call has precise breakpoints or not. Default\\n   value is true. If the call is imprecise, confidence intervals should be provided.\",\"default\":true},{\"name\":\"startWindow\",\"type\":[\"null\",\"int\"],\"doc\":\"The size of the confidence window around the start of the structural variant.\",\"default\":null},{\"name\":\"endWindow\",\"type\":[\"null\",\"int\"],\"doc\":\"The size of the confidence window around the end of the structural variant.\",\"default\":null}]}],\"doc\":\"The structural variant at this site, if the alternate allele is a structural\\n   variant. If the site is not a structural variant, this field should be left \\n   null.\",\"default\":null}]}]},{\"name\":\"annotations\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"array\",\"items\":\"string\"}},\"default\":[]},{\"name\":\"lossOfFunction\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"array\",\"items\":\"string\"}},\"default\":[]},{\"name\":\"nonsenseMediateDecay\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"array\",\"items\":\"string\"}},\"default\":[]}]}");
  public static org.apache.avro.Schema getClassSchema() { return SCHEMA$; }
  @Deprecated public org.bdgenomics.formats.avro.Variant variant;
  @Deprecated public java.util.List<java.util.List<java.lang.CharSequence>> annotations;
  @Deprecated public java.util.List<java.util.List<java.lang.CharSequence>> lossOfFunction;
  @Deprecated public java.util.List<java.util.List<java.lang.CharSequence>> nonsenseMediateDecay;

  /**
   * Default constructor.  Note that this does not initialize fields
   * to their default values from the schema.  If that is desired then
   * one should use <code>newBuilder()</code>. 
   */
  public SnpEffAnnotations() {}

  /**
   * All-args constructor.
   */
  public SnpEffAnnotations(org.bdgenomics.formats.avro.Variant variant, java.util.List<java.util.List<java.lang.CharSequence>> annotations, java.util.List<java.util.List<java.lang.CharSequence>> lossOfFunction, java.util.List<java.util.List<java.lang.CharSequence>> nonsenseMediateDecay) {
    this.variant = variant;
    this.annotations = annotations;
    this.lossOfFunction = lossOfFunction;
    this.nonsenseMediateDecay = nonsenseMediateDecay;
  }

  public org.apache.avro.Schema getSchema() { return SCHEMA$; }
  // Used by DatumWriter.  Applications should not call. 
  public java.lang.Object get(int field$) {
    switch (field$) {
    case 0: return variant;
    case 1: return annotations;
    case 2: return lossOfFunction;
    case 3: return nonsenseMediateDecay;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }
  // Used by DatumReader.  Applications should not call. 
  @SuppressWarnings(value="unchecked")
  public void put(int field$, java.lang.Object value$) {
    switch (field$) {
    case 0: variant = (org.bdgenomics.formats.avro.Variant)value$; break;
    case 1: annotations = (java.util.List<java.util.List<java.lang.CharSequence>>)value$; break;
    case 2: lossOfFunction = (java.util.List<java.util.List<java.lang.CharSequence>>)value$; break;
    case 3: nonsenseMediateDecay = (java.util.List<java.util.List<java.lang.CharSequence>>)value$; break;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }

  /**
   * Gets the value of the 'variant' field.
   */
  public org.bdgenomics.formats.avro.Variant getVariant() {
    return variant;
  }

  /**
   * Sets the value of the 'variant' field.
   * @param value the value to set.
   */
  public void setVariant(org.bdgenomics.formats.avro.Variant value) {
    this.variant = value;
  }

  /**
   * Gets the value of the 'annotations' field.
   */
  public java.util.List<java.util.List<java.lang.CharSequence>> getAnnotations() {
    return annotations;
  }

  /**
   * Sets the value of the 'annotations' field.
   * @param value the value to set.
   */
  public void setAnnotations(java.util.List<java.util.List<java.lang.CharSequence>> value) {
    this.annotations = value;
  }

  /**
   * Gets the value of the 'lossOfFunction' field.
   */
  public java.util.List<java.util.List<java.lang.CharSequence>> getLossOfFunction() {
    return lossOfFunction;
  }

  /**
   * Sets the value of the 'lossOfFunction' field.
   * @param value the value to set.
   */
  public void setLossOfFunction(java.util.List<java.util.List<java.lang.CharSequence>> value) {
    this.lossOfFunction = value;
  }

  /**
   * Gets the value of the 'nonsenseMediateDecay' field.
   */
  public java.util.List<java.util.List<java.lang.CharSequence>> getNonsenseMediateDecay() {
    return nonsenseMediateDecay;
  }

  /**
   * Sets the value of the 'nonsenseMediateDecay' field.
   * @param value the value to set.
   */
  public void setNonsenseMediateDecay(java.util.List<java.util.List<java.lang.CharSequence>> value) {
    this.nonsenseMediateDecay = value;
  }

  /** Creates a new SnpEffAnnotations RecordBuilder */
  public static org.exascience.formats.avro.SnpEffAnnotations.Builder newBuilder() {
    return new org.exascience.formats.avro.SnpEffAnnotations.Builder();
  }
  
  /** Creates a new SnpEffAnnotations RecordBuilder by copying an existing Builder */
  public static org.exascience.formats.avro.SnpEffAnnotations.Builder newBuilder(org.exascience.formats.avro.SnpEffAnnotations.Builder other) {
    return new org.exascience.formats.avro.SnpEffAnnotations.Builder(other);
  }
  
  /** Creates a new SnpEffAnnotations RecordBuilder by copying an existing SnpEffAnnotations instance */
  public static org.exascience.formats.avro.SnpEffAnnotations.Builder newBuilder(org.exascience.formats.avro.SnpEffAnnotations other) {
    return new org.exascience.formats.avro.SnpEffAnnotations.Builder(other);
  }
  
  /**
   * RecordBuilder for SnpEffAnnotations instances.
   */
  public static class Builder extends org.apache.avro.specific.SpecificRecordBuilderBase<SnpEffAnnotations>
    implements org.apache.avro.data.RecordBuilder<SnpEffAnnotations> {

    private org.bdgenomics.formats.avro.Variant variant;
    private java.util.List<java.util.List<java.lang.CharSequence>> annotations;
    private java.util.List<java.util.List<java.lang.CharSequence>> lossOfFunction;
    private java.util.List<java.util.List<java.lang.CharSequence>> nonsenseMediateDecay;

    /** Creates a new Builder */
    private Builder() {
      super(org.exascience.formats.avro.SnpEffAnnotations.SCHEMA$);
    }
    
    /** Creates a Builder by copying an existing Builder */
    private Builder(org.exascience.formats.avro.SnpEffAnnotations.Builder other) {
      super(other);
      if (isValidValue(fields()[0], other.variant)) {
        this.variant = data().deepCopy(fields()[0].schema(), other.variant);
        fieldSetFlags()[0] = true;
      }
      if (isValidValue(fields()[1], other.annotations)) {
        this.annotations = data().deepCopy(fields()[1].schema(), other.annotations);
        fieldSetFlags()[1] = true;
      }
      if (isValidValue(fields()[2], other.lossOfFunction)) {
        this.lossOfFunction = data().deepCopy(fields()[2].schema(), other.lossOfFunction);
        fieldSetFlags()[2] = true;
      }
      if (isValidValue(fields()[3], other.nonsenseMediateDecay)) {
        this.nonsenseMediateDecay = data().deepCopy(fields()[3].schema(), other.nonsenseMediateDecay);
        fieldSetFlags()[3] = true;
      }
    }
    
    /** Creates a Builder by copying an existing SnpEffAnnotations instance */
    private Builder(org.exascience.formats.avro.SnpEffAnnotations other) {
            super(org.exascience.formats.avro.SnpEffAnnotations.SCHEMA$);
      if (isValidValue(fields()[0], other.variant)) {
        this.variant = data().deepCopy(fields()[0].schema(), other.variant);
        fieldSetFlags()[0] = true;
      }
      if (isValidValue(fields()[1], other.annotations)) {
        this.annotations = data().deepCopy(fields()[1].schema(), other.annotations);
        fieldSetFlags()[1] = true;
      }
      if (isValidValue(fields()[2], other.lossOfFunction)) {
        this.lossOfFunction = data().deepCopy(fields()[2].schema(), other.lossOfFunction);
        fieldSetFlags()[2] = true;
      }
      if (isValidValue(fields()[3], other.nonsenseMediateDecay)) {
        this.nonsenseMediateDecay = data().deepCopy(fields()[3].schema(), other.nonsenseMediateDecay);
        fieldSetFlags()[3] = true;
      }
    }

    /** Gets the value of the 'variant' field */
    public org.bdgenomics.formats.avro.Variant getVariant() {
      return variant;
    }
    
    /** Sets the value of the 'variant' field */
    public org.exascience.formats.avro.SnpEffAnnotations.Builder setVariant(org.bdgenomics.formats.avro.Variant value) {
      validate(fields()[0], value);
      this.variant = value;
      fieldSetFlags()[0] = true;
      return this; 
    }
    
    /** Checks whether the 'variant' field has been set */
    public boolean hasVariant() {
      return fieldSetFlags()[0];
    }
    
    /** Clears the value of the 'variant' field */
    public org.exascience.formats.avro.SnpEffAnnotations.Builder clearVariant() {
      variant = null;
      fieldSetFlags()[0] = false;
      return this;
    }

    /** Gets the value of the 'annotations' field */
    public java.util.List<java.util.List<java.lang.CharSequence>> getAnnotations() {
      return annotations;
    }
    
    /** Sets the value of the 'annotations' field */
    public org.exascience.formats.avro.SnpEffAnnotations.Builder setAnnotations(java.util.List<java.util.List<java.lang.CharSequence>> value) {
      validate(fields()[1], value);
      this.annotations = value;
      fieldSetFlags()[1] = true;
      return this; 
    }
    
    /** Checks whether the 'annotations' field has been set */
    public boolean hasAnnotations() {
      return fieldSetFlags()[1];
    }
    
    /** Clears the value of the 'annotations' field */
    public org.exascience.formats.avro.SnpEffAnnotations.Builder clearAnnotations() {
      annotations = null;
      fieldSetFlags()[1] = false;
      return this;
    }

    /** Gets the value of the 'lossOfFunction' field */
    public java.util.List<java.util.List<java.lang.CharSequence>> getLossOfFunction() {
      return lossOfFunction;
    }
    
    /** Sets the value of the 'lossOfFunction' field */
    public org.exascience.formats.avro.SnpEffAnnotations.Builder setLossOfFunction(java.util.List<java.util.List<java.lang.CharSequence>> value) {
      validate(fields()[2], value);
      this.lossOfFunction = value;
      fieldSetFlags()[2] = true;
      return this; 
    }
    
    /** Checks whether the 'lossOfFunction' field has been set */
    public boolean hasLossOfFunction() {
      return fieldSetFlags()[2];
    }
    
    /** Clears the value of the 'lossOfFunction' field */
    public org.exascience.formats.avro.SnpEffAnnotations.Builder clearLossOfFunction() {
      lossOfFunction = null;
      fieldSetFlags()[2] = false;
      return this;
    }

    /** Gets the value of the 'nonsenseMediateDecay' field */
    public java.util.List<java.util.List<java.lang.CharSequence>> getNonsenseMediateDecay() {
      return nonsenseMediateDecay;
    }
    
    /** Sets the value of the 'nonsenseMediateDecay' field */
    public org.exascience.formats.avro.SnpEffAnnotations.Builder setNonsenseMediateDecay(java.util.List<java.util.List<java.lang.CharSequence>> value) {
      validate(fields()[3], value);
      this.nonsenseMediateDecay = value;
      fieldSetFlags()[3] = true;
      return this; 
    }
    
    /** Checks whether the 'nonsenseMediateDecay' field has been set */
    public boolean hasNonsenseMediateDecay() {
      return fieldSetFlags()[3];
    }
    
    /** Clears the value of the 'nonsenseMediateDecay' field */
    public org.exascience.formats.avro.SnpEffAnnotations.Builder clearNonsenseMediateDecay() {
      nonsenseMediateDecay = null;
      fieldSetFlags()[3] = false;
      return this;
    }

    @Override
    public SnpEffAnnotations build() {
      try {
        SnpEffAnnotations record = new SnpEffAnnotations();
        record.variant = fieldSetFlags()[0] ? this.variant : (org.bdgenomics.formats.avro.Variant) defaultValue(fields()[0]);
        record.annotations = fieldSetFlags()[1] ? this.annotations : (java.util.List<java.util.List<java.lang.CharSequence>>) defaultValue(fields()[1]);
        record.lossOfFunction = fieldSetFlags()[2] ? this.lossOfFunction : (java.util.List<java.util.List<java.lang.CharSequence>>) defaultValue(fields()[2]);
        record.nonsenseMediateDecay = fieldSetFlags()[3] ? this.nonsenseMediateDecay : (java.util.List<java.util.List<java.lang.CharSequence>>) defaultValue(fields()[3]);
        return record;
      } catch (Exception e) {
        throw new org.apache.avro.AvroRuntimeException(e);
      }
    }
  }
}
